File: physics/muon/muon_c/lsq.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

void lsq_data(int size, int badbins, double *bin_center,double *bin_freq) {
    
    double x, y, sxlny, slny, sx2, sx, a, tau;
    sxlny = 0.0;
    slny = 0.0;
    sx2 = 0.0;
    sx = 0.0;
    for(int i = badbins; i < size; i++){
        x = bin_center[i];
        y = bin_freq[i];
        if((int)bin_freq[i]==0){y = 1.0;}
        sxlny += x*log(y);
        slny += log(y);
        sx2 += x*x;
        sx += x;
    }
    a = exp((slny*sx2 - sx*sxlny)/(size*sx2 - sx*sx));
    tau = -1.0/((size*sxlny - sx*slny)/(size*sx2 - sx*sx));
    fprintf(stderr,"%f  %f\n", a, tau);
}

void wlsq_data(int size, int badbins, double *bin_center,double *bin_freq, double *a, double *tau) {
    
    double x, y, sxlny, sx2y, sxy, slny, sx2, sx, sy, sylny, sxylny, denom;
    sxlny = 0.0;
    sx2y = 0.0;
    sxy = 0.0;
    slny = 0.0;
    sx2 = 0.0;
    sx = 0.0;
    sy = 0.0;
    sylny = 0.0;
    sxylny = 0.0;
    for(int i = badbins; i < size; i++){
        x = bin_center[i];
        y = bin_freq[i];
        if((int)bin_freq[i]==0){y = 1.0;}
        sxlny += x*log(y);
        sx2y += x*x*y;
        sxy += x*y;
        slny += log(y);
        sx2 += x*x;
        sx += x;
        sy += y;
        sylny += y*log(y);
        sxylny += x*y*log(y);
    }
    denom = sy*sx2y - sxy*sxy;
    *a = exp((sx2y*sylny - sxy*sxylny)/denom);
    *tau = -1.0/((-sxy*sylny + sy*sxylny)/denom);
    fprintf(stderr, "%f  %f\n", *a, *tau);
}

void calculate_background(int size, int last,double *bin_freq, double *c) {
    double count = 0.0;
    for(int i = size-1; i >= size-(last+1); i--){
        count += bin_freq[i];
    }
    *c = count/(double) last;
    fprintf(stderr, "%f\n", *c);
}

File: physics/mfp/lsq.js
function Lsq(x, y) {
    this.x = x;
    this.y = y;
    this.lsq = function() {
        var sxlny, slny, sx2, sx, n;
        sxlny = 0;
        slny = 0;
        sx2 = 0;
        sx = 0;
        n = this.x.length
        for (var i = 0; i < x.length; i++) {
            if (y[i] == 0) y[i] = 1;
            sxlny += x[i]*Math.log(y[i]);
            slny += Math.log(y[i]);
            sx2 += x[i]*x[i];
            sx += x[i];
        }
        return [Math.exp((slny*sx2 - sx*sxlny)/(n*sx2 - sx*sx)), (n*sxlny - sx*slny)/(n*sx2 - sx*sx)]
    }

    this.wlsq = function() {
        var sxlny, sx2y, sxy, slny, sx2, sx, sy, sylny, sxylny;
        sxlny = 0.0;
        sx2y = 0.0;
        sxy = 0.0;
        slny = 0.0;
        sx2 = 0.0;
        sx = 0.0;
        sy = 0.0;
        sylny = 0.0;
        sxylny = 0.0;
        for (var i = 0; i < x.length; i++) {
            if (y[i] == 0) y[i] = 1;
            sxlny += x[i]*Math.log(y[i]);
            sx2y += x[i]*x[i]*y[i];
            sxy += x[i]*y[i];
            slny += Math.log(y[i]);
            sx2 += x[i]*x[i];
            sx += x[i];
            sy += y[i];
            sylny += y[i]*Math.log(y[i]);
            sxylny += x[i]*y[i]*Math.log(y[i]);
        }
        return [Math.exp((sx2y*sylny - sxy*sxylny)/(sy*sx2y-sxy*sxy)),
                   (-(sxy*sylny)+sy*sxylny)/(sy*sx2y - (sxy*sxy))];
    }
}